apps/presale/contracts/BalanceRedirectPresale.sol

```diff
@@ -0,0 +1,235 @@
+        require(_value > 0,                                                        ERROR_INVALID_CONTRIBUTE_VALUE);
+        require(contributionToken.balanceOf(_contributor) >= _value,               ERROR_INSUFFICIENT_BALANCE);
+        require(contributionToken.allowance(_contributor, address(this)) >= _value, ERROR_INSUFFICIENT_ALLOWANCE);
+
+        // (contributor) ~~~> contribution tokens ~~~> (presale)
+        _transfer(contributionToken, _contributor, address(this), _value);
+
+        // (mint ✨) ~~~> project tokens ~~~> (contributor)
+        uint256 tokensToSell = contributionToTokens(_value);
+        tokenManager.mint(_contributor, tokensToSell);
+        totalRaised = totalRaised.add(_value);
+
+        emit Contribute(_contributor, _value, tokensToSell);
+    }
+
+    /**
+     * @notice Does nothing. Interface compliance.
+    */
+    function refund(address, uint256) external isInitialized {
+        return;
+    }
+
+    /**
+     * @notice Close presale and open trading
+    */
+    function close() external nonReentrant isInitialized {
+        require(state() == State.Funding, ERROR_INVALID_STATE);
+
+        isClosed = true;
+
+        // (presale) ~~~> contribution tokens ~~~> (reserve)
+        uint256 tokensForReserve = totalRaised.mul(futureReserveRatio) / PPM;
+        _transfer(contributionToken, address(this), reserve, tokensForReserve);
+
+        // (presale) ~~~> contribution tokens ~~~> (beneficiary)
+        uint256 fundsForBeneficiary = contributionToken.balanceOf(address(this));
+        if (fundsForBeneficiary > 0) {
+            _transfer(contributionToken, address(this), beneficiary, fundsForBeneficiary);
+        }
+
+        // open trading
+        controller.openTrading();
+
+        emit Close();
+    }
+
+    /***** public view functions *****/
+
+    /**
+     * @notice Computes the amount of [bonded] tokens that would be purchased for `@tokenAmount(self.contributionToken(): address, _value)`
+     * @param _value The amount of contribution tokens to be used in that computation
+    */
+    function contributionToTokens(uint256 _value) public view isInitialized returns (uint256) {
+        return _value.mul(exchangeRate).div(PPM);
+    }
+
+    /**
+     * @notice Returns the current state of that presale
+    */
+    function state() public view isInitialized returns (State) {
+        if (openDate == 0 || openDate > getTimestamp64()) {
+            return State.Pending;
+        }
+
+        if (isClosed || _timeSinceOpen() >= period) {
+            return State.Closed;
+        }
+
+        return State.Funding;
+    }
+
+    /***** internal functions *****/
+
+    function _timeSinceOpen() internal view returns (uint64) {
+        if (openDate == 0) {
+            return 0;
+        } else {
+            return getTimestamp64().sub(openDate);
+        }
+    }
+
+    function _setOpenDate(uint64 _date) internal {
```

Comment:
Would be interesting if we could expose a way to change or simply delay this one in case we need to delay the start time of the Court (see https://github.com/aragon/aragon-court/pull/211)